libparsec/crates/client/src/certif/device.rs

```diff
@@ -0,0 +1,175 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::sync::Arc;
+
+use libparsec_client_connection::{AuthenticatedCmds, ConnectionError};
+use libparsec_platform_storage::certificates::GetCertificateError;
+use libparsec_protocol::authenticated_cmds::v4::device_create;
+use libparsec_types::{
+    anyhow, thiserror, Bytes, CertificateSignerOwned, DateTime, DeviceCertificate, LocalDevice,
+    MaybeRedacted, SigningKeyAlgorithm, UserID,
+};
+
+use super::{
+    encrypt::CertifEncryptForUserError, greater_timestamp, CertificateBasedActionOutcome,
+    GreaterTimestampOffset, InvalidCertificateError,
+};
+
+pub async fn register_new_device(
+    cmds: Arc<AuthenticatedCmds>,
+    new_device: LocalDevice,
+    author: Arc<LocalDevice>,
+) -> Result<CertificateBasedActionOutcome, CertifDeviceError> {
+    // Loop is needed to deal with server requiring greater timestamp
+    let mut timestamp = author.now();
+
+    loop {
+        let outcome = internal_register_new_device(
+            cmds.clone(),
+            new_device.clone(),
+            author.clone(),
+            timestamp,
+        )
+        .await?;
+
+        match outcome {
+            DeviceInternalsOutcome::Done(outcome) => return Ok(outcome),
+            DeviceInternalsOutcome::RequireGreaterTimestamp(strictly_greater_than) => {
+                // TODO: handle `strictly_greater_than` out of the client ballpark by
+                // returning an error
+                timestamp = greater_timestamp(
+                    &author.time_provider,
+                    GreaterTimestampOffset::User,
+                    strictly_greater_than,
+                );
+            }
+        }
+    }
+}
+
+pub(crate) struct DeviceCertificatesBytes {
+    pub full: Bytes,
+    pub redacted: Bytes,
+}
+
+/// generates certificates for new device signed by author at now
+pub(crate) fn generate_new_device_certificates(
+    new_device: &LocalDevice,
+    author: Arc<LocalDevice>,
+    now: DateTime,
+) -> DeviceCertificatesBytes {
+    let device_cert = DeviceCertificate {
+        author: CertificateSignerOwned::User(author.device_id),
+        timestamp: now,
+        user_id: new_device.user_id,
+        device_id: new_device.device_id,
+        device_label: MaybeRedacted::Real(new_device.device_label.clone()),
+        verify_key: new_device.verify_key(),
+        algorithm: SigningKeyAlgorithm::Ed25519,
+    };
+
+    let device_certificate = device_cert.dump_and_sign(&author.signing_key).into();
+
+    let redacted_device_cert = device_cert.into_redacted();
+
+    let redacted_device_certificate = redacted_device_cert
+        .dump_and_sign(&author.signing_key)
+        .into();
+
+    DeviceCertificatesBytes {
+        full: device_certificate,
+        redacted: redacted_device_certificate,
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CertifDeviceError {
```

Comment:
The function using this error is called `register_new_device`, so it should be renamed `CertifRegisterNewDeviceError`